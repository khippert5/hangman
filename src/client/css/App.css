
html, body {
  margin: 0;
  padding: 0;
}

body * {
  font-size: 16px;
  margin: 0;
  padding: 0;
  position: relative;
  z-index: 1;
  box-sizing: border-box;
}
