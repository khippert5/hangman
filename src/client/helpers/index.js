// @flow

// Actions
import { letterSelect, setErrorMessage } from '../lib/actions';

// Constants
import MESSAGES from '../constants/messages';

// Helprs

// Helper to check of the work has all the letters in the validLetters array
// return { Boolean }
export const checkSuccess = (words: Array<string>, validLetters: Array<string>) => {
  const captureGuesses = [];
  words.map((w) => {
    const wordMap = w.split('');
    for (let i = 0; i < wordMap.length; i++) {
      if (validLetters.indexOf(wordMap[i]) !== -1) {
        captureGuesses.push(wordMap[i]);
      }
    }
    return w;
  });

  const isGuess = captureGuesses.join('');
  return Boolean(isGuess === words.join(''));
};

// Helper to get random message from message constant file
// return { string }
export const getMessage = (message: string, key?: string): string => {
  if (!message) return 'Something went wrong';

  let isError = MESSAGES[message];
  isError = isError[Math.floor(Math.random() * (isError.length - 0 + 1)) + 0] || isError[0];
  if (key) {
    return isError.replace('[letterReplace]', key.toUpperCase());
  }
  return isError;
};

// Function run when added to the window.addEventListner and window.removeEventListner
// return { void }
export const setKeyPressHandler = (e: { key: string }) => {
  const { key } = e;
  if (/[a-z]/gi.test(key)) {
    try {
      if (window.hangman.ui && window.hangman.ui.selected.includes(key)) {
        window.hangman.dispatch(setErrorMessage('duplicate', key));
        return;
      }
      window.hangman.dispatch(letterSelect(key));
    } catch (err) {
      console.log('Dispatch not found');
      console.log(err);
    }
  }
};

// Helper to allow addition of keypress listener to window
// return { void }
export const registerWindowKeyPress = (): void => {
  document.addEventListener('keypress', setKeyPressHandler);
};

// Helper to allow removal of keypress listener from window
// return { void }
export const removeWindowKeyPress = (): void => {
  document.removeEventListener('keypress', setKeyPressHandler);
};


export default registerWindowKeyPress;
