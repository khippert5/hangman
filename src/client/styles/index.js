// @flow

import styled from 'styled-components';

import COLORS from './configs/colors';
import RULES from './configs/rules';
import BREAKPOINTS from '../constants/breakpoints';

export const AppWrapper = styled.div`
  padding: 16px;
  @media screen (max-width: ${BREAKPOINTS.phone}px) {
    padding: 6px;
  }
`;

export const AppContainer = styled.div`
  max-width: 500px;
  margin: 0 auto;
  width: 100%;

  * {
    line-height: 1.45em;
  }
`;

export const DevTag = styled.p`
  font-size: 0.9em;
  float: right;
  margin: 20px 0;
  span {
    font-weight: 600;
  }
`;

export const FooterContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-content: center;
  > * {
    width: 100%;
    margin-right: 15px;
    &:last-child {
      margin-right: 0;
      text-align: right;
    }
  }

  @media screen and (max-width:${BREAKPOINTS.phone}px) {
    flex-direction: column;
  }
`;

export const GuessLine = styled.p`
  font-size: 1.25em;
  font-weight: 600;

  span {
    color: ${(props: { color: string }) => COLORS[props.color]}
    font-size: inherit;
  }
`;

export const Letter = styled.button`
  background: ${(props: { isSelected: boolean}) => (props.isSelected ? COLORS.disabledGrey : '#FFF')};
  border: 1px solid ${(props: { isSelected: boolean}) => (props.isSelected ? COLORS.disabledGrey : COLORS.orange)};
  border-radius: 3px;
  ${(props: { isSelected: boolean}) => !props.isSelected && 'box-shadow: 1px 1px 4px #CCC;'}
  color: ${(props: { isSelected: boolean}) => (props.isSelected ? COLORS.disabledText : COLORS.orange)};
  font-weight: 600;
  padding: 5px;
  margin: 10px 5px 5px 0;
  width: 33px;
  text-transform: uppercase;
  &:focus, &:hover {
    outline: none;
  }
  ${(props: { isSelected: boolean}) => (!props.isSelected && `
    &:hover {
      background: ${COLORS.orange};
      color: #FFF;
      box-shadow: initial;
      cursor: pointer;
    }`
  )}
`;

export const LetterSelectContainer = styled.div`
  align: center;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  width: 100%;
`;

export const Message = styled.p`
  font-size: 1em;
  margin-bottom: ${RULES.marginBottom};
  margin-top: 15px;
  ${(props: { messageStyle: string }) => (props.messageStyle === 'error' && `color: ${COLORS.error};`)}
  ${(props: { messageStyle: string }) => (props.messageStyle === 'correct' && `color: ${COLORS.green};`)}
  ${(props: { messageStyle: string }) => (props.messageStyle !== 'correct' && props.messageStyle !== 'error' && `color: ${COLORS.black};`)}
  font-weight: 600;
`;

export const SubTitle = styled.h3`
    font-size: 1.75em;
    font-weight: 400;
    margin-bottom: ${RULES.marginBottom};
`;

export const Title = styled.h2`
  font-size: 2.75em;
  margin-bottom: ${RULES.marginBottom};
`;
