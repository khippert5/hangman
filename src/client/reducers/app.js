// @flow

// Helpers
import { registerWindowKeyPress } from '../helpers';

// Types
import type { AppState } from '../lib/types/reducer-states';
import type { Action } from '../lib/types/actions';

export const defaultState: AppState = {
  init: {
    date: null,
    time: null,
  },
};

export default function (state: AppState = defaultState, action: Action): AppState {
  switch (action.type) {
    case 'INIT_APPLICATION': {
      registerWindowKeyPress(); // Register keypress to window for keyboard input
      const newState = state;
      // Set date time started
      const date = new Date();
      const newDate = `${date.getFullYear()}/${date.getDate()}/${date.getMonth() + 1}`;
      const time = `${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}:${date.getMilliseconds()}`;

      return {
        ...newState,
        init: {
          date: newDate,
          time,
        },
      };
    }
    default:
      return state;
  }
}
