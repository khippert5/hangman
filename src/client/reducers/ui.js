// @flow

// Helpers
import {
  checkSuccess, getMessage, removeWindowKeyPress, registerWindowKeyPress,
} from '../helpers';
import { getWord, getWordPosition } from './helpers/ui';

// Types
import type { UiState } from '../lib/types/reducer-states';
import type { Action } from '../lib/types/actions';

export const defaultState: UiState = {
  correct: null,
  error: null,
  incorrect: 4,
  selected: [],
  // eslint-disable-next-line no-use-before-define
  solution: getWord(),
  success: false,
  validLetters: [],
  values: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'],
};

export default function (state: UiState = defaultState, action: Action): UiState {
  switch (action.type) {
    case 'RESET_GAME': {
      let newState = state;
      // Clean keypress from window so it does not duplicate
      removeWindowKeyPress();
      // Re-register keypress to window
      registerWindowKeyPress();

      newState = {
        ...newState,
        correct: null,
        error: null,
        incorrect: 4,
        selected: [],
        success: false,
        validLetters: [],
      };

      // Register to window for keypress
      // Needed to pull from on the keypress register event
      window.hangman = {
        ...window.hangman,
        ui: {
          ...newState,
        },
      };

      return newState;
    }
    case 'SET_ERROR_MESSAGE': {
      const { message, key } = action;
      let newState = state;
      newState = {
        ...newState,
        error: getMessage(message, key),
      };

      // Register to window for keypress
      window.hangman = {
        ...window.hangman,
        ui: {
          ...newState,
        },
      };

      return newState;
    }
    case 'SET_SELECTED_LETTER': {
      let newState = state;
      const { letter } = action;
      const {
        incorrect, selected, validLetters, solution,
      } = newState;
      let success = false;
      const newSelected = selected;
      if (!solution) return newState;
      const { word } = solution;

      // Failout for keypress
      if (selected.includes(letter)) return state;

      // Remove keypress listener from winow to ensure it doesn't continue to trigger if failed
      removeWindowKeyPress();

      let newIncorrect = incorrect;
      let error = null;
      let correct = getMessage('correct', letter);
      if (word.indexOf(letter) === -1) {
        newIncorrect++;
        error = getMessage('wrong', letter);
        correct = null;
      }
      if (word.indexOf(letter) !== -1) {
        validLetters.push(letter);
      }

      newSelected.push(letter);

      if (word !== null) {
        success = checkSuccess(word.split(' '), validLetters);
        if (success) correct = getMessage('success', letter);
      }

      // If not met limit or not successful, re-register keypress to allow futher input
      if (!success && newIncorrect < 10) registerWindowKeyPress();

      newState = {
        ...newState,
        correct,
        error,
        incorrect: newIncorrect,
        selected: newSelected,
        success,
        validLetters,
      };

      // Register to window for keypress
      window.hangman = {
        ...window.hangman,
        ui: {
          ...newState,
        },
      };

      // Return computed state
      return newState;
    }
    case 'START_NEW_GAME': {
      let newState = state;
      const { solution } = newState;
      const newSolution = getWord(solution, getWordPosition(solution));
      // Clean keypress
      removeWindowKeyPress();
      // Re-register keypress
      registerWindowKeyPress();

      newState = {
        ...newState,
        correct: null,
        error: null,
        incorrect: 4,
        selected: [],
        solution: newSolution,
        success: false,
        validLetters: [],
      };

      // Register to window for keypress
      window.hangman = {
        ...window.hangman,
        ui: {
          ...newState,
        },
      };

      return newState;
    }
    default:
      return state;
  }
}
