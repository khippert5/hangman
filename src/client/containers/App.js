// @flow

import React, { Fragment } from 'react';
import { useDispatch, useSelector } from 'react-redux';

// Components
import HangmanComponent from '../components/Hangman';
import LetterSelect from '../components/fragments/Select-letters';
import WordPreview from '../components/fragments/Word';
import MessageContainer from '../components/fragments/Message';
import GuessContainer from '../components/fragments/Guesses';
import ErrorBoundary from './Error-boundary';
import Failure from '../components/fragments/Failure';
import FooterWrapper from '../components/fragments/Footer';
import SuccessWrapper from '../components/fragments/Success';

// Styles
import '../css/App.css';
import { WordContainer } from '../styles/Word';

// Actions
import { initApplication } from '../lib/actions';

// Styles
import {
  AppWrapper, AppContainer, Title,
} from '../styles';

// Types
import type { UiState } from '../lib/types/reducer-states';
import type { AppProps } from '../lib/types';

export const App = () => {
  // const [state] = useReducer(Reducer);
  const dispatch = useDispatch();

  // Set to window for use later
  window.hangman = window.hangman || {};
  window.hangman.dispatch = dispatch;

  const init = useSelector((state) => state.app.init);
  const {
    incorrect, success,
  }: AppProps = useSelector((state): UiState => state.ui);
  const stop = incorrect > 9;

  // Init app is not set
  if (!init || (init && init.time === null)) {
    dispatch(initApplication(dispatch));
  }

  // Error boundary used to catch thrown messages to ensure no blank screen and some user interaction.
  // Further functions would be used to correct the application depending on the error. Great feature for local dev to catch things before production
  return (<ErrorBoundary>
    <AppWrapper>
      <AppContainer>
        <Title>React Hangman</Title>
          <HangmanComponent incorrectGuessCount={incorrect}></HangmanComponent>
          <WordContainer>
            {success && !stop && <SuccessWrapper />}
            <WordPreview isSuccess={success} />
          </WordContainer>
          {!success && <Fragment>
            <GuessContainer />
            {!stop && <MessageContainer />}
          </Fragment>}
          {!success && stop && <Failure />}
          {!success && !stop && <LetterSelect /> }
        <FooterWrapper />
      </AppContainer>
    </AppWrapper>
  </ErrorBoundary>);
};

export default App;
