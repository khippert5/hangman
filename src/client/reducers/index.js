// @flow

import { combineReducers } from 'redux';
import type { Reducer } from 'redux';

import app, { defaultState as appState } from './app';
import ui, { defaultState as uiState } from './ui';

import type { State } from '../lib/types/reducer-states';
import type { Action } from '../lib/types/actions';

const reducer: Reducer<State, Action> = combineReducers({
  app,
  ui,
});

export default reducer;

export const defaultState: State = {
  app: appState,
  ui: uiState,
};
