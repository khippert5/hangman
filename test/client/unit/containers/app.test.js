import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import {canvas as mockCanvas } from 'canvas';

// State
import { defaultState } from '../../../../src/client/reducers';

// Container
import App from '../../../../src/client/containers/App';

// Styles
import Button from '../../../../src/client/styles/buttons';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const identify = jest.fn();
const dispatch = jest.fn();
window.analytics = { identify };
window.HTMLCanvasElement.prototype.getContext = jest.fn();
global.console = {
  ...global.console,
  warn: (val) => {},
  error: (val) => {},
}

describe('Test application render', () => {
  let wrapper;
  const newState = defaultState;

  it('It renders successfully', async (done) => {
    const setState = newState;
    setState.ui.solution = {
      word: 'test',
      hint: 'Test solution',
    };
    const store = mockStore(setState);
    const component = renderer
      .create(<Provider store={store}><App store={store} /></Provider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    done();
  });
  it('Test no error on button click', async (done) => {
    let setState = newState;
    setState.ui.solution = {
      word: 'test',
      hint: 'Test solution',
    };
    setState.ui.incorrect = 5
    let store = mockStore(setState);
    wrapper = mount(<Provider store={store}><App store={store} /></Provider>);
    wrapper.find('button.new').simulate('click');
    wrapper.find('button.reset').simulate('click');

    setState = newState;
    setState.ui.solution = {
      word: 'test',
      hint: 'Test solution',
    };
    setState.ui.success = true;
    store = mockStore(setState);
    wrapper = mount(<Provider store={store}><App store={store} /></Provider>);
    wrapper.find('button.new').simulate('click');

    setState = newState;
    setState.ui.solution = {
      word: 'test',
      hint: 'Test solution',
    };
    setState.ui.success = null;
    setState.ui.incorrect = 10;
    store = mockStore(setState);
    wrapper = mount(<Provider store={store}><App store={store} /></Provider>);
    wrapper.find('button.new').simulate('click');
    wrapper.find('button.reset').simulate('click');

    done();
  });
});
