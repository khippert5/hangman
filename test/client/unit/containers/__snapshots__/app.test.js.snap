// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test application render It renders successfully 1`] = `
<div
  className="sc-bdVaJa cwHqWd"
>
  <div
    className="sc-bwzfXH krYIhT"
  >
    <h2
      className="sc-dnqmqq jZWzqX"
    >
      React Hangman
    </h2>
    <div
      className="Hangman"
    >
      <canvas />
    </div>
    <div
      className="sc-iwsKbI hOKQXu"
    >
      <p>
        <b>
          Hint:
        </b>
         
        Test solution
      </p>
      <h3>
        Your word is:
      </h3>
      <div
        className="sc-VigVT eiRcZc"
      >
        <div
          className="sc-gZMcBi iEOUPw"
        >
          <div
            className="sc-gqjmRU iyjRIq"
          >
             
          </div>
          <div
            className="sc-gqjmRU iyjRIq"
          >
             
          </div>
          <div
            className="sc-gqjmRU iyjRIq"
          >
             
          </div>
          <div
            className="sc-gqjmRU iyjRIq"
          >
             
          </div>
        </div>
      </div>
    </div>
    <p
      className="sc-ifAKCX kdopqP"
      color="green"
    >
      You have 
      <span>
        6
      </span>
       incorrect tries to get it right.
    </p>
    <p
      className="sc-gzVnrw hsqWdA"
    >
      Use your keyboard or click on a letter to get started
    </p>
    <button
      className="sc-EHOje bXraum"
      data-value="a"
      onClick={[Function]}
    >
      a
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="b"
      onClick={[Function]}
    >
      b
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="c"
      onClick={[Function]}
    >
      c
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="d"
      onClick={[Function]}
    >
      d
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="e"
      onClick={[Function]}
    >
      e
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="f"
      onClick={[Function]}
    >
      f
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="g"
      onClick={[Function]}
    >
      g
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="h"
      onClick={[Function]}
    >
      h
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="i"
      onClick={[Function]}
    >
      i
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="j"
      onClick={[Function]}
    >
      j
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="k"
      onClick={[Function]}
    >
      k
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="l"
      onClick={[Function]}
    >
      l
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="m"
      onClick={[Function]}
    >
      m
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="n"
      onClick={[Function]}
    >
      n
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="o"
      onClick={[Function]}
    >
      o
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="p"
      onClick={[Function]}
    >
      p
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="q"
      onClick={[Function]}
    >
      q
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="r"
      onClick={[Function]}
    >
      r
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="s"
      onClick={[Function]}
    >
      s
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="t"
      onClick={[Function]}
    >
      t
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="u"
      onClick={[Function]}
    >
      u
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="v"
      onClick={[Function]}
    >
      v
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="w"
      onClick={[Function]}
    >
      w
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="x"
      onClick={[Function]}
    >
      x
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="y"
      onClick={[Function]}
    >
      y
    </button>
    <button
      className="sc-EHOje bXraum"
      data-value="z"
      onClick={[Function]}
    >
      z
    </button>
    <div
      className="sc-bxivhb geuGCY"
    >
      <div
        className="sc-fjdhpX cQrmnA"
      >
        <button
          className="sc-jTzLTM rHsMs new"
          onClick={[Function]}
        >
          New puzzle
        </button>
      </div>
      <p
        className="sc-htpNat bOVmHZ"
      >
        <span>
          Developer:
        </span>
         Kevin Hippert
      </p>
    </div>
  </div>
</div>
`;
